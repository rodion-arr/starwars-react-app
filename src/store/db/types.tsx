import {Char} from "./entities/char.d";
import {Film} from "./entities/film.d";
import {Planet} from "./entities/planet.d";
import {Specie} from "./entities/specie.d";
import {Starship} from "./entities/starship.d";
import {Timeline} from "./entities/timeline";
import {Vehicle} from "./entities/vehicle.d";

export const types = Object.freeze({
  GET_DB: '[DB] GET_DB',
  FILL_PEOPLE: '[DB] FILL_PEOPLE',
  FILL_FILMS: '[DB] FILL_FILMS',
  FILL_PLANETS: '[DB] FILL_PLANETS',
  FILL_SPECIES: '[DB] FILL_SPECIES',
  FILL_STARSHIPS: '[DB] FILL_STARSHIPS',
  FILL_VEHICLES: '[DB] FILL_VEHICLES',
  FILL_TIMELINE: '[DB] FILL_TIMELINE',
});

export interface DbState {
  people: Record<string, Char> | null;
  films: Record<string, Film> | null;
  planets: Record<string, Planet> | null;
  species: Record<string, Specie> | null;
  starships: Record<string, Starship> | null;
  vehicles: Record<string, Vehicle> | null;
  timeline: Record<string, Timeline> | null;
}

export type DbObject = Char | Film | Planet | Specie | Starship | Vehicle | Timeline;
