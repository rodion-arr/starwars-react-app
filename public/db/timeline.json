{"1":{"id":"1","title":"The Phantom Menace","start":-31,"end":-31},"2":{"id":"2","title":"Attack of the Clones","start":-22,"end":-22},"3":{"id":"3","title":"The Clone Wars (film)","start":-22,"end":-22},"4":{"id":"4","title":"The Clone Wars: Season One","start":-22,"end":-21},"5":{"id":"5","title":"The Clone Wars: Season Two","start":-21,"end":-21},"6":{"id":"6","title":"The Clone Wars: Season Three","start":-22,"end":-20},"7":{"id":"7","title":"The Clone Wars: Season Four","start":-20,"end":-19},"8":{"id":"8","title":"The Clone Wars: Season Seven","start":-19,"end":-19},"9":{"id":"9","title":"Star Wars: Revenge of the Sith","start":-19,"end":-19},"10":{"id":"10","title":"The Bad Batch","start":-19,"end":-19},"11":{"id":"11","title":"Solo: A Star Wars Story","start":-13,"end":-10},"12":{"id":"12","title":"Obi-Wan Kenobi","start":-9,"end":-9},"13":{"id":"13","title":"Rebels: Season One","start":-5,"end":-4},"14":{"id":"14","title":"Rebels: Season Two","start":-4,"end":-3},"15":{"id":"15","title":"Rebels: Season Three","start":-2,"end":-2},"16":{"id":"16","title":"Rebels: Season Four","start":-1,"end":0},"17":{"id":"17","title":"Andor","start":-5,"end":0},"18":{"id":"18","title":"Rogue One: A Star Wars Story","start":0,"end":0},"19":{"id":"19","title":"A New Hope","start":0,"end":0},"20":{"id":"20","title":"The Empire Strikes Back","start":3,"end":3},"21":{"id":"21","title":"Return of the Jedi","start":4,"end":4},"22":{"id":"22","title":"The Mandalorian","start":9,"end":9},"23":{"id":"23","title":"Ahsoka","start":9,"end":9},"24":{"id":"24","title":"Resistance: Season One","start":34,"end":34},"25":{"id":"25","title":"Resistance: Season Two","start":34,"end":35},"26":{"id":"26","title":"The Force Awakens","start":34,"end":34},"27":{"id":"27","title":"The Last Jedi","start":34,"end":34},"28":{"id":"28","title":"The Rise of Skywalker","start":35,"end":34},"29":{"id":"29","title":"The Book of Boba Fett","start":9,"end":9}}
