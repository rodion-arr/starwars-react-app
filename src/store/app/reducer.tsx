import {AppState} from "./types";
import {createSlice} from "@reduxjs/toolkit";

const initialState: AppState = {
};

export const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {},
  extraReducers: (builder) => {

  },
});
