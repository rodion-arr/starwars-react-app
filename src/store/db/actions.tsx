import {createAction, createAsyncThunk} from "@reduxjs/toolkit";
import {Char} from "./entities/char.d";
import {types} from "./types";
import {AppDispatch} from "../store";
import {DbService} from "../../services/db.service";
import {Film} from "./entities/film.d";
import {Planet} from "./entities/planet.d";
import {Specie} from "./entities/specie.d";
import {Starship} from "./entities/starship.d";

export const dbActions = Object.freeze({
  fillPeople: createAction<Record<string, Char> | null>(types.FILL_PEOPLE),
  fillFilms: createAction<Record<string, Film> | null>(types.FILL_FILMS),
  fillPlanets: createAction<Record<string, Planet> | null>(types.FILL_PLANETS),
  fillSpecies: createAction<Record<string, Specie> | null>(types.FILL_SPECIES),
  fillStarships: createAction<Record<string, Starship> | null>(types.FILL_STARSHIPS),

  // Async
  getDb: createAsyncThunk<
    void,
    void,
    {
      dispatch: AppDispatch;
    }
    >(types.GET_DB, async (arg, { dispatch}) => {
      dispatch(dbActions.fillPeople(await DbService.loadDbFile('/db/people.json')))
      dispatch(dbActions.fillFilms(await DbService.loadDbFile('/db/films.json')))
      dispatch(dbActions.fillPlanets(await DbService.loadDbFile('/db/planets.json')))
      dispatch(dbActions.fillSpecies(await DbService.loadDbFile('/db/species.json')))
      dispatch(dbActions.fillStarships(await DbService.loadDbFile('/db/starships.json')))
  }),
});
