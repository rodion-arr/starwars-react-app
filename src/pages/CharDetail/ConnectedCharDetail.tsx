import { connect } from 'react-redux';
import { CharDetail } from './CharDetail';
import {AppState} from "../../store/root.reducer";

const mapState = (state: AppState) => ({
  people: state.db.people,
  planets: state.db.planets,
  films: state.db.films,
  species: state.db.starships,
  vehicles: state.db.vehicles,
  starships: state.db.starships,
});

const mapDispatch = {};

export const connector = connect(mapState, mapDispatch);

const ConnectedCharDetail = connect(mapState, mapDispatch)(CharDetail);
export default ConnectedCharDetail;
