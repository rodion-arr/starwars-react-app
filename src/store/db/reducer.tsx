import {DbState} from "./types";
import {createSlice, PayloadAction} from "@reduxjs/toolkit";
import {dbActions} from "./actions";
import {Char} from "./entities/char.d";
import {Film} from "./entities/film.d";
import {Planet} from "./entities/planet.d";
import {Specie} from "./entities/specie.d";
import {Starship} from "./entities/starship.d";

const initialState: DbState = {
  people: null,
  films: null,
  planets: null,
  species: null,
  starships: null,
};

export const dbSlice = createSlice({
  name: 'db',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      dbActions.fillPeople.type,
      (state, action: PayloadAction<Record<string, Char>>) => {
        state.people = action.payload;
      },
    );

    builder.addCase(
      dbActions.fillFilms.type,
      (state, action: PayloadAction<Record<string, Film>>) => {
        state.films = action.payload;
      },
    );

    builder.addCase(
      dbActions.fillPlanets.type,
      (state, action: PayloadAction<Record<string, Planet>>) => {
        state.planets = action.payload;
      },
    );

    builder.addCase(
      dbActions.fillSpecies.type,
      (state, action: PayloadAction<Record<string, Specie>>) => {
        state.species = action.payload;
      },
    );

    builder.addCase(
      dbActions.fillStarships.type,
      (state, action: PayloadAction<Record<string, Starship>>) => {
        state.starships = action.payload;
      },
    );
  },
});
