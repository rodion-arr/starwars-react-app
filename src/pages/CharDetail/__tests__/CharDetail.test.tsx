import {configureStore} from "@reduxjs/toolkit";
import {db} from "../../../../tests/fixtures/db";
import {render} from "@testing-library/react";
import {Provider} from "react-redux";
import * as reactRouter from "react-router";
import { BrowserRouter as Router } from "react-router-dom"
import {rootReducer} from "../../../store/root.reducer";
import ConnectedCharDetail from "../ConnectedCharDetail";

describe('CharDetail', () => {
  it('should render correctly', function () {
    jest.spyOn(reactRouter, 'useParams').mockImplementationOnce(() => ({id: '1'}));

    const storeMock = configureStore({
      reducer: rootReducer,
      preloadedState: db
    });

    const {container} = render(
      <Provider store={storeMock}>
        <Router>
          <ConnectedCharDetail />
        </Router>
      </Provider>);

    expect(container).toMatchSnapshot();
  });

  it('should not render if no data', function () {
    jest.spyOn(reactRouter, 'useParams').mockImplementationOnce(() => ({id: '1'}));

    const storeMock = configureStore({
      reducer: rootReducer,
      preloadedState: {}
    });

    const {container} = render(
      <Provider store={storeMock}>
        <Router>
          <ConnectedCharDetail />
        </Router>
      </Provider>);

    expect(container).toMatchSnapshot();
  });
});
