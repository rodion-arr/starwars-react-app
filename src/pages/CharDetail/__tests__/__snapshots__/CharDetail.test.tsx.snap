// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharDetail should not render if no data 1`] = `<div />`;

exports[`CharDetail should render correctly 1`] = `
<div>
  <div
    class="entity-detail"
  >
    <div
      class="entity-detail__title-wrap"
    >
      <h2
        class="entity-detail__title"
      >
        Luke Skywalker
      </h2>
    </div>
    <div
      class="entity-detail__img-wrap"
    >
      <img
        alt="Luke Skywalker"
        class="entity-detail__img"
        src="/assets/content/people/1.png"
      />
    </div>
    <div
      class="entity-detail__title-wrap"
    >
      <h2
        class="entity-detail__title"
      >
        Characteristics
      </h2>
    </div>
    <div
      class="entity-detail__spec detail-spec-table"
    >
      <table
        class="detail-spec-table__table"
      >
        <tbody>
          <tr>
            <td>
              Height
            </td>
            <td>
              172
            </td>
          </tr>
          <tr>
            <td>
              Mass
            </td>
            <td>
              77
            </td>
          </tr>
          <tr>
            <td>
              Hair color
            </td>
            <td>
              blond
            </td>
          </tr>
          <tr>
            <td>
              Skin color
            </td>
            <td>
              fair
            </td>
          </tr>
          <tr>
            <td>
              Eye color
            </td>
            <td>
              blue
            </td>
          </tr>
          <tr>
            <td>
              Birth year
            </td>
            <td>
              19BBY
            </td>
          </tr>
          <tr>
            <td>
              Gender
            </td>
            <td>
              male
            </td>
          </tr>
          <tr>
            <td>
              Home world
            </td>
            <td>
              <a
                href="/planets/1"
              >
                Tatooine
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="catalog"
    >
      <h2
        class="catalog__title"
      >
        Films
      </h2>
      <div
        class="catalog__items catalog-items-wrap"
      >
        <div
          class="catalog-item"
        >
          <a
            class="catalog-item__link"
            href="/films/1"
          >
            <div
              class="catalog-item__img-wrap"
            >
              <img
                alt="A New Hope"
                class="catalog-item__img"
                loading="lazy"
                src="/assets/content/films/1.png"
              />
            </div>
            <div
              class="catalog-item__info catalog-item-info"
            >
              <div
                class="catalog-item-info__name"
              >
                A New Hope
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
    <div
      class="catalog"
    >
      <h2
        class="catalog__title"
      >
        Species
      </h2>
      <div
        class="catalog__items catalog-items-wrap"
      >
        <div
          class="catalog-item"
        >
          <a
            class="catalog-item__link"
            href="/species/1"
          >
            <div
              class="catalog-item__img-wrap"
            >
              <img
                alt="Human"
                class="catalog-item__img"
                loading="lazy"
                src="/assets/content/species/1.png"
              />
            </div>
            <div
              class="catalog-item__info catalog-item-info"
            >
              <div
                class="catalog-item-info__name"
              >
                Human
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
    <div
      class="catalog"
    >
      <h2
        class="catalog__title"
      >
        Vehicles
      </h2>
      <div
        class="catalog__items catalog-items-wrap"
      >
        <div
          class="catalog-item"
        >
          <a
            class="catalog-item__link"
            href="/vehicles/1"
          >
            <div
              class="catalog-item__img-wrap"
            >
              <img
                alt="Sand Crawler"
                class="catalog-item__img"
                loading="lazy"
                src="/assets/content/vehicles/1.png"
              />
            </div>
            <div
              class="catalog-item__info catalog-item-info"
            >
              <div
                class="catalog-item-info__name"
              >
                Sand Crawler
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
    <div
      class="catalog"
    >
      <h2
        class="catalog__title"
      >
        Starships
      </h2>
      <div
        class="catalog__items catalog-items-wrap"
      >
        <div
          class="catalog-item"
        >
          <a
            class="catalog-item__link"
            href="/starships/1"
          >
            <div
              class="catalog-item__img-wrap"
            >
              <img
                alt="CR90 corvette"
                class="catalog-item__img"
                loading="lazy"
                src="/assets/content/starships/1.png"
              />
            </div>
            <div
              class="catalog-item__info catalog-item-info"
            >
              <div
                class="catalog-item-info__name"
              >
                CR90 corvette
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
